Docker/Runner/Limits.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+namespace Keboola\DockerBundle\Docker\Runner;
+
+use Keboola\DockerBundle\Docker\Image;
+use Keboola\Syrup\Exception\ApplicationException;
+use Psr\Log\LoggerInterface;
+
+class Limits
+{
+    /**
+     * @var array
+     */
+    private $userFeatures;
+
+    /**
+     * @var array
+     */
+    private $projectFeatures;
+
+    /**
+     * @var array
+     */
+    private $projectLimits;
+
+    /**
+     * @var array
+     */
+    private $instanceLimits;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * Limits constructor.
+     * @param LoggerInterface $logger
+     * @param array $instanceLimits
+     * @param array $projectLimits
+     * @param array $projectFeatures
+     * @param array $userFeatures
+     */
+    public function __construct(
```

Comment:
co kdyby v konstruktoru byl rovnou ten image? limity kontejneru jsou průnikem featur (uživatele, projektu, komponenty) a limitů (uživatelem, projektu, komponenty a instance). všechno by se tam mělo předat na začátku.

případně by se v ctoru mohly předat rovnou `componentFeatures` a `componentLimits`

Reply:
kdyby mel byt image v ctoru, tak by se musel limits objekt vytvaret v cyklu tady https://github.com/keboola/docker-bundle/blob/odin-container-limits/Service/Runner.php#L407-L419
coz me bylo proti srsti v tom smyslu, ze ty limity jsou pevne dane v okamziku spusteni a jejich inicializace v cyklu zavani tim, ze by se snad mohly menit

da se do ctoru predat pole imagu (coz jsem tak puvodne mel), ale ve vysledku tam k nicemu neni, protoze `Container` se stejne pta na limity konkretniho image. Tzn, musel by se proparovat podle ID a je to uplne zbytecny, protoze container ten image, ktery spousti, zna.

Reply:
Jasně, už tomu rozumím. `Limits` jsi pojal jako nějakej obecnej storage (perzistentní?) pro featury a limity, kterej na požádání pro danej image zjistí nějaký nastavení. Já se na to díval mnohem užším pohledem - pro danej kontejner bych tam nasypal featury (usera, projektu, komponenty), nastavení limitů projektu, komponent a instance, vznikl by z toho immutable, kde by sis jenom načetl, jak máš nastavit CPU pro danej běh kontejneru. Pro novej kontejner by se to celý zopakovalo.

Pokud maj všechny metody stejný argument, tak bych ho asi strčil do konstruktoru?

Reply:
vsak to je immutable ne? 

> Pokud maj všechny metody stejný argument, tak bych ho asi strčil do konstruktoru?

to nechapu

Reply:
jasně, je. v každý metodě máš argument `Image $image`, takže očividně výsledek záleží na $image. v jednom cyklu runneru (kontejner) tam bude furt stejný `$image`, tak jsem se to jen snažil trochu normalizovat :-).

Reply:
aha, no tak to je to co jsem psal https://github.com/keboola/docker-bundle/pull/288#discussion_r185804187 - musel by se ten limits vytvaret v kazde iteraci tech images, tak jako muze to tak byt - na funkci to nema vliv - jen me to proste prijde, ze to pak vypada, ze ty limity jsou jine pro kazdy image

Reply:
no tak pak to bude limit komponenty a timpadem bude soucasti `$image`, obecny limity (instance, projekt, user) budou furt stejny

Reply:
takže ten jeden projekt bude mít teď 8 jader na všechny komponenty? co kdybychom ještě chtěli říct, že některý komponenty jsou striktně jednojádrový, tak to bude 
`min($instance, $project, $component)` a v `$component` bude maximální počet cpu nastavených pro komponentu?

Reply:
jo a jo. Ted bych to resil tak, ze ten jeden projekt bude mit 8 jader pro vsechno. Dal bych to asi probral s padakem a martinem jaky jsou mozny usecasy, podle toho v tom potom zavedl nejakej system a pravidla, at nepridavame zmatene limity ke komponentam a projektum

Reply:
Ok, v tom případě ✅ .

Reply:
pokracovani https://github.com/keboola/internal/issues/1